staging/src/k8s.io/apiserver/pkg/cel/template/template.go

```diff
@@ -0,0 +1,193 @@
+	envSet := environment.MustBaseEnvSet(ver)
+
+	objectOpt := cel.Variable(ObjectName, types.MapType)
+	versionedObjectOpt := environment.VersionedOptions{
+		IntroducedVersion: version.MajorMinor(1, 29), // TODO: not correct (?), but necessary to use it with environment.DefaultCompatibilityVersion() = 1.29.
+		EnvOptions:        []cel.EnvOption{objectOpt},
+	}
+	envSet, err := envSet.Extend(versionedObjectOpt)
+	if err != nil {
+		return nil, fmt.Errorf("extending CEL environment with object variable: %w", err)
+	}
+
+	c := config{
+		envType: environment.NewExpressions,
+	}
+	for _, opt := range options {
+		if err := opt(&c); err != nil {
+			return nil, err
+		}
+	}
+	env, err := envSet.Env(c.envType)
+	if err != nil {
+		return nil, fmt.Errorf("create CEL environment: %w", err)
+	}
+
+	t := &Template{
+		env:      env,
+		template: template,
+	}
+
+	var errs []error
+	indices := placeholderRe.FindAllStringIndex(template, -1)
+	t.expressions = make([]expression, len(indices))
+	for i, offsets := range indices {
+		t.expressions[i].start = offsets[0]
+		t.expressions[i].end = offsets[1]
+
+		// TODO: provide a custom common.Location which handles mapping offset back to the full template.
+		ast, issues := env.CompileSource(common.NewStringSource(template[offsets[0]+2:offsets[1]-2], fmt.Sprintf("placeholder at #%d", offsets[0])))
+		if issues != nil && issues.Err() != nil {
+			errs = append(errs, fmt.Errorf("CEL type-check error: %w", issues.Err()))
+			continue
+		}
+		program, err := env.Program(ast)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("CEL program construction error: %w", err))
+			continue
+		}
+		t.expressions[i].program = program
+	}
+	if len(errs) > 0 {
+		return nil, errors.Join(errs...)
+	}
+
+	return t, nil
+}
+
+func EnvironmentType(envType environment.Type) Option {
+	return func(c *config) error {
+		c.envType = envType
+		return nil
+	}
+}
+
+type Option func(c *config) error
+
+type config struct {
+	envType environment.Type
+}
+
+type Template struct {
+	env         *cel.Env
+	template    string
+	expressions []expression
+}
+
+type expression struct {
+	// Start and  end character (exclusive) of the {{...}} in the template string.
+	start, end int
+
+	program cel.Program
+}
+
+func (t *Template) Expand(ctx context.Context, object map[string]any) (string, error) {
+	if len(t.expressions) == 0 {
+		return t.template, nil
+	}
+
+	replacements := make([]string, 0, len(t.expressions))
+	lenReplaced := 0
+	lenReplacements := 0
+
+	value := types.NewStringInterfaceMap(t.env.CELTypeAdapter(), object)
+
+	var errs []error
+	for _, expr := range t.expressions {
+		lenReplaced += expr.end - expr.start
+		replacementVal, _, err := expr.program.ContextEval(ctx,
+			map[string]any{
+				ObjectName: value,
+			},
+		)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("CEL evaluation error: %w", err))
+			continue
+		}
+		replacementAny, err := replacementVal.ConvertToNative(stringType)
```

Comment:
So this is indeed where it now fails when the expression is of some non-string type.

For the purpose of DRA, it could be specified that the `{{ ... }}` gets converted to `any` here and then encoded as JSON. That works because YAML is a true superset of JSON.

It's required also for strings, because without quoting it isn't save to just insert the input string into a string which contains YAML.